### PR TITLE
fix: pod workers test

### LINKED ISSUE #105589: TestUpdatePodWithTerminatedPod flaking
### Which jobs are flaking?

pull-kubernetes-unit

### Which tests are flaking?

TestUpdatePodWithTerminatedPod

### Since when has it been flaking?

10/8 (when it was introduced in https://github.com/kubernetes/kubernetes/pull/105527)

### Testgrid link

_No response_

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=TestUpdatePodWithTerminatedPod

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit&exclude-non-failed-tests=50&width=20

Running with [stress](https://pkg.go.dev/golang.org/x/tools/cmd/stress) shows immediate failures:

```
go test -race -c ./pkg/kubelet
stress ./kubelet.test -test.v -test.run TestUpdatePodWithTerminatedPod
```

### Relevant SIG(s)

/sig node

### PR DESCRIPTION
**Pull request by jonyhy96**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Running pod update with type `kubetypes.SyncPodKill` will eventually become terminated.Because it is handled in seperate goroutines, the main test goroutine may not perceive the latest changes of pod status.

#### Which issue(s) this PR fixes:

Fixes #105589

### CODE REVIEW

**Comment by jonyhy96** [comment]
PTAL @liggitt

**Comment by 249043822** [comment]
Yea, I thinks this change may make sense or we can even delete this test, need @ehashman or @rphillips decision, thanks

**Review by ehashman — CHANGES_REQUESTED** [review]
/hold

I don't think this is the right fix.

**Comment by jonyhy96** [comment]
ping @ehashman @liggitt @249043822 should we push this forward?

**Comment by pacoxu** [comment]
TestUpdatePodWithTerminatedPod still flakes https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103967/pull-kubernetes-unit/1470317029310337024

Any update？

**Comment by cynepco3hahue** [comment]
nope, will try to take a look this week

**Comment by cynepco3hahue** [comment]
Can you please answer @ehashman comment?

**Comment by jonyhy96** [comment]
> Can you please answer @ehashman comment?

see https://github.com/kubernetes/kubernetes/pull/105599#discussion_r729470730

**Comment by jonyhy96** [comment]
/unhold 
ping @ehashman @cynepco3hahue could we push this forward?

**Comment by jonyhy96** [comment]
ping @sjenning @klueska @dims PTAL

File: pkg/kubelet/pod_workers_test.go
Comment (tallclair) [reply pkg/kubelet/pod_workers_test.go:313]: Yeah, the `RunningPod` field is only populated by the kubelet when cleaning up orphaned containers:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1118-L1134

I'm not sure what the original intent was for this test, but terminating the orphaned pod represented by this is expected. It looks like the running pod case is already covered by the variable `pod`, so I think maybe this case can just be deleted?

Comment (jonyhy96) [reply pkg/kubelet/pod_workers_test.go:313]: runningPod and pod are different type, maybe the original intent was test podWorkers works for both type of pod. But the other promblem here is the original test didn't call 	drainAllWorkers(podWorkers) befor assert pod status.

File: pkg/kubelet/pod_workers_test.go
Comment (tallclair) [inline pkg/kubelet/pod_workers_test.go:314]:
```suggestion
	if podWorkers.IsPodKnownTerminated(orphanedPod.ID) == false {
		t.Errorf("podWorker state should be terminated for orphaned pod")
```

File: pkg/kubelet/pod_workers_test.go
Comment (tallclair) [reply pkg/kubelet/pod_workers_test.go:313]: Right, I guess if we want to leave the test case in place, then it's expected that it's terminated, which is what you have here. It might help clear things up to rename `runningPod` to `orphanedPod` (suggestion below).

**Comment by jonyhy96** [comment]
already updated, PTAL @tallclair

**Review by tallclair — COMMENTED** [review]
/lgtm
/approve

**Comment by liggitt** [comment]
echoing https://github.com/kubernetes/kubernetes/pull/105599#issuecomment-942334352, I lack context on the code under test, so I don't know if the change is correct or not. I'll defer to https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/OWNERS for approval

**Comment by rphillips** [comment]
lgtm as well

**Comment by ehashman** [comment]
/lgtm
as well -- thanks for the updates.

**Comment by pacoxu** [comment]
milestone label is needed.
https://github.com/kubernetes/kubernetes/issues/105589 is marked to be fixed in v1.24.

**Comment by dims** [comment]
There are lgtm(s) from @ehashman @tallclair and @rphillips 

/milestone v1.24

**Comment by dims** [comment]
/approve

LGTM as well.

**Comment by liggitt** [comment]
filed https://github.com/kubernetes/kubernetes/issues/109258 for integration flake

**Pull request merged**
